File: sfpp/pinsketch.py
"""PinSketch over GF(2^32), ported from src/modules/Native/PinSketch.cpp.

This is a transcription, not a reimplementation: it exists so the simulator's sketches are the
same bytes the firmware would put on the air. Where the two could drift, oracle.py compiles the
C++ and compares, so a divergence fails a test rather than quietly changing a result.

Field: GF(2^32) modulo x^32 + x^7 + x^3 + x^2 + 1.

The caches below are memoisation of pure functions, not approximations: every byte this module
produces is unchanged, which check_oracle.py verifies against the firmware's own PinSketch.cpp.
"""

from functools import lru_cache

MODULUS = 0x8D
FIELD_BITS = 32
MASK = 0xFFFFFFFF


def mul(a, b):
    r = 0
    while b:
        if b & 1:
            r ^= a
        b >>= 1
        overflow = a & 0x80000000
        a = (a << 1) & MASK
        if overflow:
            a ^= MODULUS
    return r


def sqr(a):
    return mul(a, a)


# 62 multiplications a call, and the polynomial routines below ask for the same handful of
# leading coefficients over and over: measured 13863 hits against 281 misses in a single run.
@lru_cache(maxsize=1 << 16)
def inv(a):
    # a^(2^32-2), the exponent being the sum of 2^1..2^31.
    r, p = 1, a
    for _ in range(1, FIELD_BITS):
        p = sqr(p)
        r = mul(r, p)
    return r


# --- Polynomials over GF(2^32), coefficient index == degree, no trailing zeros ---


def _trim(p):
    while p and p[-1] == 0:
        p.pop()
    return p


def _deg(p):
    return len(p) - 1


def _add_to(a, b):
    if len(a) < len(b):
        a.extend([0] * (len(b) - len(a)))
    for i, coef in enumerate(b):
        a[i] ^= coef
    return _trim(a)


def _mul_poly(a, b):
    if not a or not b:
        return []
    r = [0] * (len(a) + len(b) - 1)
    for i, ai in enumerate(a):
        if ai == 0:
            continue
        for j, bj in enumerate(b):
            r[i + j] ^= mul(ai, bj)
    return _trim(r)


def _divmod_poly(num, den, want_quotient=False):
    """Long division. Returns (remainder, quotient); quotient is [] unless asked for."""
    if not den:
        return [], []
    rem = _trim(list(num))
    dd = _deg(den)
    # A monic divisor is the common case here, and needs no inverse at all.
    lead_inv = 1 if den[-1] == 1 else inv(den[-1])
    quotient = [0] * (_deg(rem) - dd + 1) if want_quotient and _deg(rem) >= dd else []

    while rem and _deg(rem) >= dd:
        shift = _deg(rem) - dd
        coef = mul(rem[-1], lead_inv)
        if want_quotient:
            quotient[shift] = coef
        for i in range(dd + 1):
            rem[shift + i] ^= mul(coef, den[i])
        _trim(rem)
    return rem, _trim(quotient)


def _gcd_poly(a, b):
    a, b = _trim(list(a)), _trim(list(b))
    while b:
        r, _ = _divmod_poly(a, b)
        a, b = b, r
    return a


def _trace_poly(beta, f):
    """Tr(beta*z) reduced mod f, where Tr(y) = y + y^2 + y^4 + ... + y^(2^31)."""
    h, _ = _divmod_poly([0, beta], f)
    acc = list(h)
    for _ in range(1, FIELD_BITS):
        h, _ = _divmod_poly(_mul_poly(h, h), f)
        _add_to(acc, h)
    return acc


def _find_roots(f, out):
    """Berlekamp trace algorithm. False if f does not split completely over the field."""
    deg = _deg(f)
    if deg <= 0:
        return (
            deg == 0
        )  # nonzero constant has no roots; the zero polynomial is not decodable
    if deg == 1:
        out.append(mul(f[0], inv(f[1])))
        return True
    for k in range(FIELD_BITS):
        g = _gcd_poly(f, _trace_poly(1 << k, f))
        dg = _deg(g)
        if 0 < dg < deg:
            _, q = _divmod_poly(f, g, want_quotient=True)
            return _find_roots(g, out) and _find_roots(q, out)
    return False


def _berlekamp_massey(s):
    """Shortest connection polynomial over the syndrome sequence."""
    c, b = [1], [1]
    l, m = 0, 1
    bb = 1

    for n in range(len(s)):
        d = s[n]
        for i in range(1, min(l, len(c) - 1) + 1):
            d ^= mul(c[i], s[n - i])
        if d == 0:
            m += 1
            continue
        prev = list(c)
        coef = mul(d, inv(bb))
        if len(c) < len(b) + m:
            c.extend([0] * (len(b) + m - len(c)))
        for i, bi in enumerate(b):
            c[i + m] ^= mul(coef, bi)
        if 2 * l <= n:
            l = n + 1 - l
            b = prev
            bb = d
            m = 1
        else:
            m += 1
    return _trim(c)


# The same short ID is added by every server that hears it, and again by decode()'s verification:
# measured 1584 adds over 98 distinct (element, capacity) pairs in one run.
@lru_cache(maxsize=1 << 16)
def _odd_powers(e, capacity):
    """e, e^3, e^5 ... e^(2*capacity-1). Returned shared and read-only - never mutate it."""
    step = sqr(e)
    out = []
    power = e
    for _ in range(capacity):
        out.append(power)
        power = mul(power, step)
    return tuple(out)


class Sketch:
    """A fixed-size digest holding the odd power sums of its members."""

    def __init__(self, capacity=0):
        self.syndromes = [0] * capacity

    @property
    def capacity(self):
        return len(self.syndromes)

    def serialized_size(self):
        return len(self.syndromes) * 4

    def add(self, e):
        """Toggles membership - adding a held element removes it again. Rejects zero."""
        if e == 0:
            return False
        syndromes = self.syndromes
        for i, power in enumerate(_odd_powers(e, len(syndromes))):
            syndromes[i] ^= power
        return True

    def merge(self, other):
        """XORs an equal-capacity sketch in, leaving the sketch of the symmetric difference."""
        if len(other.syndromes) != len(self.syndromes):
            return False
        for i in range(len(self.syndromes)):
            self.syndromes[i] ^= other.syndromes[i]
        return True

    def empty(self):
        return all(s == 0 for s in self.syndromes)

    def clear(self):
        self.syndromes = [0] * len(self.syndromes)

    def truncate(self, new_capacity):
        """Discards capacity from the tail. Only ever shrinks - what makes prefix streaming work."""
        if new_capacity < len(self.syndromes):
            self.syndromes = self.syndromes[:new_capacity]

    def copy(self):
        s = Sketch(0)
        s.syndromes = list(self.syndromes)
        return s

    def serialize(self):
        out = bytearray()
        for e in self.syndromes:
            out += bytes(
                ((e) & 0xFF, (e >> 8) & 0xFF, (e >> 16) & 0xFF, (e >> 24) & 0xFF)
            )
        return bytes(out)

    def deserialize(self, data):
        if len(data) % 4 != 0:
            return False
        self.syndromes = [
            int.from_bytes(data[i : i + 4], "little") for i in range(0, len(data), 4)
        ]
        return True

    def decode(self):
        """Recovers the set, sorted ascending, or None when it is not decodable.

        Subject to the ~1/c! misdecode rate: an over-capacity difference lands on a wrong set that
        reproduces the same syndromes at about that rate.
        """
        if not self.syndromes:
            return None
        if self.empty():
            return []

        # Odd syndromes are carried; even ones are free, since S_2j == S_j^2 in characteristic 2.
        c = len(self.syndromes)
        s = [0] * (2 * c)
        for j in range(1, 2 * c + 1):
            s[j - 1] = self.syndromes[(j - 1) // 2] if j & 1 else sqr(s[j // 2 - 1])

        locator = _berlekamp_massey(s)
        if not locator or locator[0] != 1 or _deg(locator) > c:
            return None

        roots = []
        if not _find_roots(locator, roots) or len(roots) != _deg(locator):
            return None

        found = []
        for r in roots:
            if r == 0:
                return None
            found.append(inv(r))
        found.sort()
        if any(found[i] == found[i + 1] for i in range(len(found) - 1)):
            return None

        # The locator is only a candidate until the recovered set reproduces the syndromes we
        # started from. Without this an over-capacity difference decodes to a plausible wrong set.
        check = Sketch(c)
        for e in found:
            check.add(e)
        if check.syndromes != self.syndromes:
            return None
        return found

File: sfpp/native/PinSketch.cpp
#include "PinSketch.h"

#include <algorithm>

namespace pinsketch
{

// Low-order terms of x^32 + x^7 + x^3 + x^2 + 1.
static constexpr Element MODULUS = 0x8D;
static constexpr int FIELD_BITS = 32;

Element mul(Element a, Element b)
{
    Element r = 0;
    while (b) {
        if (b & 1)
            r ^= a;
        b >>= 1;
        const bool overflow = a & 0x80000000u;
        a <<= 1;
        if (overflow)
            a ^= MODULUS;
    }
    return r;
}

Element sqr(Element a)
{
    return mul(a, a);
}

Element inv(Element a)
{
    // a^(2^32-2), the exponent being the sum of 2^1..2^31.
    Element r = 1, p = a;
    for (int i = 1; i < FIELD_BITS; i++) {
        p = sqr(p);
        r = mul(r, p);
    }
    return r;
}

// --- Polynomials over GF(2^32), coefficient index == degree, no trailing zeros ---

using Poly = std::vector<Element>;

static void polyTrim(Poly &p)
{
    while (!p.empty() && p.back() == 0)
        p.pop_back();
}

// Degree of the zero polynomial is reported as -1.
static int polyDeg(const Poly &p)
{
    return (int)p.size() - 1;
}

static void polyAddTo(Poly &a, const Poly &b)
{
    if (a.size() < b.size())
        a.resize(b.size(), 0);
    for (size_t i = 0; i < b.size(); i++)
        a[i] ^= b[i];
    polyTrim(a);
}

static Poly polyMul(const Poly &a, const Poly &b)
{
    if (a.empty() || b.empty())
        return Poly();
    Poly r(a.size() + b.size() - 1, 0);
    for (size_t i = 0; i < a.size(); i++) {
        if (a[i] == 0)
            continue;
        for (size_t j = 0; j < b.size(); j++)
            r[i + j] ^= mul(a[i], b[j]);
    }
    polyTrim(r);
    return r;
}

// Long division, returning the remainder and optionally the quotient.
static Poly polyDivMod(const Poly &num, const Poly &den, Poly *quotient = nullptr)
{
    if (quotient)
        quotient->clear();
    if (den.empty())
        return Poly();

    Poly rem = num;
    polyTrim(rem);
    const int dd = polyDeg(den);
    const Element leadInv = inv(den.back());
    if (quotient && polyDeg(rem) >= dd)
        quotient->assign(polyDeg(rem) - dd + 1, 0);

    while (polyDeg(rem) >= dd && !rem.empty()) {
        const int shift = polyDeg(rem) - dd;
        const Element coef = mul(rem.back(), leadInv);
        if (quotient)
            (*quotient)[shift] = coef;
        for (int i = 0; i <= dd; i++)
            rem[shift + i] ^= mul(coef, den[i]);
        polyTrim(rem);
    }
    if (quotient)
        polyTrim(*quotient);
    return rem;
}

static Poly polyGcd(Poly a, Poly b)
{
    polyTrim(a);
    polyTrim(b);
    while (!b.empty()) {
        Poly r = polyDivMod(a, b);
        a = b;
        b = r;
    }
    return a;
}

// Tr(beta*z) reduced mod f, where Tr(y) = y + y^2 + y^4 + ... + y^(2^31).
static Poly tracePoly(Element beta, const Poly &f)
{
    Poly h = polyDivMod(Poly{0, beta}, f);
    Poly acc = h;
    for (int i = 1; i < FIELD_BITS; i++) {
        h = polyDivMod(polyMul(h, h), f);
        polyAddTo(acc, h);
    }
    return acc;
}

/**
 * Berlekamp trace algorithm. Splits f by Tr(beta*x), which separates any two distinct roots for at
 * least one basis element beta, then recurses on both factors. Returns false if f does not split
 * completely over the field, which is how an invalid locator is caught.
 */
static bool findRoots(const Poly &f, std::vector<Element> &out)
{
    const int deg = polyDeg(f);
    if (deg <= 0)
        return deg == 0; // a nonzero constant has no roots; the zero polynomial is not decodable
    if (deg == 1) {
        out.push_back(mul(f[0], inv(f[1])));
        return true;
    }
    for (int k = 0; k < FIELD_BITS; k++) {
        const Poly g = polyGcd(f, tracePoly((Element)1u << k, f));
        const int dg = polyDeg(g);
        if (dg > 0 && dg < deg) {
            Poly q;
            polyDivMod(f, g, &q);
            return findRoots(g, out) && findRoots(q, out);
        }
    }
    return false;
}

/**
 * Berlekamp-Massey over the syndrome sequence, yielding the shortest connection polynomial. For a
 * difference set of size d <= capacity that polynomial is the error locator, whose roots are the
 * inverses of the set members.
 */
static Poly berlekampMassey(const std::vector<Element> &s)
{
    Poly c{1}, b{1};
    size_t l = 0, m = 1;
    Element bb = 1;

    for (size_t n = 0; n < s.size(); n++) {
        Element d = s[n];
        for (size_t i = 1; i <= l && i < c.size(); i++)
            d ^= mul(c[i], s[n - i]);
        if (d == 0) {
            m++;
            continue;
        }
        const Poly prev = c;
        const Element coef = mul(d, inv(bb));
        if (c.size() < b.size() + m)
            c.resize(b.size() + m, 0);
        for (size_t i = 0; i < b.size(); i++)
            c[i + m] ^= mul(coef, b[i]);
        if (2 * l <= n) {
            l = n + 1 - l;
            b = prev;
            bb = d;
            m = 1;
        } else {
            m++;
        }
    }
    polyTrim(c);
    return c;
}

// --- Sketch ---

bool Sketch::add(Element e)
{
    if (e == 0)
        return false;
    const Element step = sqr(e);
    Element power = e;
    for (size_t i = 0; i < syndromes.size(); i++) {
        syndromes[i] ^= power;
        power = mul(power, step);
    }
    return true;
}

bool Sketch::merge(const Sketch &other)
{
    if (other.syndromes.size() != syndromes.size())
        return false;
    for (size_t i = 0; i < syndromes.size(); i++)
        syndromes[i] ^= other.syndromes[i];
    return true;
}

bool Sketch::empty() const
{
    for (Element s : syndromes)
        if (s != 0)
            return false;
    return true;
}

void Sketch::clear()
{
    std::fill(syndromes.begin(), syndromes.end(), 0);
}

void Sketch::truncate(size_t newCapacity)
{
    if (newCapacity < syndromes.size())
        syndromes.resize(newCapacity);
}

void Sketch::serialize(uint8_t *out) const
{
    for (size_t i = 0; i < syndromes.size(); i++) {
        const Element e = syndromes[i];
        out[i * 4 + 0] = (uint8_t)(e & 0xFF);
        out[i * 4 + 1] = (uint8_t)((e >> 8) & 0xFF);
        out[i * 4 + 2] = (uint8_t)((e >> 16) & 0xFF);
        out[i * 4 + 3] = (uint8_t)((e >> 24) & 0xFF);
    }
}

bool Sketch::deserialize(const uint8_t *in, size_t len)
{
    if (len % sizeof(Element) != 0)
        return false;
    syndromes.assign(len / sizeof(Element), 0);
    for (size_t i = 0; i < syndromes.size(); i++)
        syndromes[i] = (Element)in[i * 4 + 0] | ((Element)in[i * 4 + 1] << 8) | ((Element)in[i * 4 + 2] << 16) |
                       ((Element)in[i * 4 + 3] << 24);
    return true;
}

bool Sketch::decode(std::vector<Element> &out) const
{
    out.clear();
    if (syndromes.empty())
        return false;
    if (empty())
        return true;

    // Odd syndromes are carried; even ones are free, since S_2j == S_j^2 in characteristic 2.
    const size_t c = syndromes.size();
    std::vector<Element> s(2 * c, 0);
    for (size_t j = 1; j <= 2 * c; j++)
        s[j - 1] = (j & 1) ? syndromes[(j - 1) / 2] : sqr(s[j / 2 - 1]);

    const Poly locator = berlekampMassey(s);
    if (locator.empty() || locator[0] != 1 || (size_t)polyDeg(locator) > c)
        return false;

    std::vector<Element> roots;
    if (!findRoots(locator, roots) || roots.size() != (size_t)polyDeg(locator))
        return false;

    std::vector<Element> found;
    found.reserve(roots.size());
    for (Element r : roots) {
        if (r == 0)
            return false;
        found.push_back(inv(r));
    }
    std::sort(found.begin(), found.end());
    if (std::adjacent_find(found.begin(), found.end()) != found.end())
        return false;

    // The locator is only a candidate until the recovered set reproduces the syndromes we started
    // from. Without this a difference larger than the capacity can decode to a plausible wrong set.
    Sketch check(c);
    for (Element e : found)
        if (!check.add(e))
            return false;
    if (check.syndromes != syndromes)
        return false;

    out = std::move(found);
    return true;
}

} // namespace pinsketch
